File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;
        int nwStart=newInterval[0], nwEnd=newInterval[1], len=intervals.size(), i=0;

// if start is MORE than end of interval, then add all
        while(i < len && nwStart > intervals[i][1]){
            res.push_back(intervals[i]);
            i++;
        }

// lies in the range   nwStart >= intervals[i][0] && nwStart <= intervals[i][1]
        while(i < len && nwEnd >= intervals[i][0]){
                nwStart = min(nwStart, intervals[i][0]);
                nwEnd = max(nwEnd, intervals[i][1]);
                i++;
        }

        res.push_back({nwStart, nwEnd});

        // if remaining
        while(i < len)
            res.push_back(intervals[i++]);
        
        return res;
    }
};

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
  public int[][] insert(int[][] intervals, int[] newInterval) {

    int len=intervals.length, newStart=newInterval[0], newEnd=newInterval[1], i=0;
    List<int[]> res = new ArrayList<>();

    // push element inside the new res till start is MORE than interval end
    while (i < len) {
      if (newStart > intervals[i][1]) // if start is MORE than interval end
        res.add(intervals[i]);
      else if (newEnd < intervals[i][0]) // if end LESS than interval start
        break;
      else {
        // Overlap : merge them
        newStart = Math.min(newStart, intervals[i][0]);
        newEnd = Math.max(newEnd, intervals[i][1]);
      }
      i++;
    }

    res.add(new int[] {newStart,newEnd});

    while (i < len)
      res.add(intervals[i++]);
    
    return res.toArray(new int[res.size()][2]);
  }
}